pkg/os/volume/api.go

```diff
@@ -268,6 +272,96 @@ func getTarget(mount string) (string, error) {
 	return volumeString, nil
 }
 
+// GetVolumeIDFromTargetPath returns the volume id of a given target path.
+func (VolumeAPI) GetClosestVolumeIDFromTargetPath(targetPath string) (string, error) {
+	volumeString, err := findClosestVolume(targetPath)
+
+	if err != nil {
+		return "", fmt.Errorf("error getting the closest volume for the path=%s, err=%v", targetPath, err)
+	}
+
+	return volumeString, nil
+}
+
+// findClosestVolume finds the closest volume id for a given target path
+// by following symlinks and moving up in the filesystem, if after moving up in the filesystem
+// we get to a DriveLetter then the volume corresponding to this drive letter is returned instead.
+func findClosestVolume(path string) (string, error) {
+	candidatePath := path
+
+	// run in a bounded loop to avoid doing an infinite loop
+	// while trying to follow symlinks
+	for i := 0; i < 256; i += 1 {
+		fi, err := os.Lstat(candidatePath)
+		if err != nil {
+			return "", err
+		}
+		isSymlink := fi.Mode()&os.ModeSymlink != 0
+
+		if isSymlink {
+			target, err := dereferenceSymlink(candidatePath)
+			if err != nil {
+				return "", err
+			}
+			// if it has the form Volume{volumeid}\ then it's a volume
+			if strings.HasPrefix(target, "Volume") {
```

Comment:
Will it be possible to make this a more rigorous check? For example, a regex to match the Volume{volumeID} pattern rather than a prefix and also checking with `Get-Volume` when the regex matches.

Reply:
@jingxu97 shared some code from k/k that uses this function: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/subpath/subpath_windows.go#L62:6, what I saw is that after dereferencing a symlink it had the form: `Volume{X}` or `C:/foo/bar`